### ISSUE TITLE
Research: Implement "Fast KV Compaction via Attention Matching" for extreme VRAM savings

### ISSUE THREAD

**Issue by Gavin0725** [ggml-org/llama.cpp#20037]
### Research Stage

- [x] Background Research (Let's try to avoid reinventing the wheel)
- [x] Hypothesis Formed (How do you think this will work and it's effect?)
- [ ] Strategy / Implementation Forming
- [ ] Analysis of results
- [ ] Debrief / Documentation (So people in the future can learn from us)

### Previous existing literature and research

As context lengths grow, KV cache VRAM consumption becomes a massive bottleneck for local inference on consumer hardware. Current workarounds like prompt summarization or token dropping are lossy and degrade model quality. Other methods like the "Cartridges" approach require end-to-end gradient optimization, which is far too slow and computationally expensive for on-the-fly local inference.
​A newly published paper (Feb 18, 2026), "Fast KV Compaction via Attention Matching" (https://arxiv.org/abs/2602.16284), proposes a breakthrough solution to this problem.

### Hypothesis

The paper introduces a method to compress the K and V matrices into much smaller representations that stay in the latent space. The model directly queries against these compact matrices during generation without needing to decompress them.
​Expected Effects:
​Massive VRAM Savings: Up to 50x KV cache compression with near-lossless generation quality.
​Lightning Fast: The compression step takes only seconds because it uses closed-form solutions (e.g., Moore-Penrose pseudoinverse) to match attention outputs, entirely avoiding slow backpropagation.
​Faster Decoding: Since there is no decompression step and the attention matrices are physically smaller, generation speed could potentially increase.

### Implementation

I am opening this as an RFC (Request for Comments) to put this on the core team's radar. I do not have a PR ready yet.
​The main technical hurdle for llama.cpp appears to be the math involved. The algorithm requires calculating the Moore-Penrose pseudoinverse of a matrix to solve an overdetermined system (least-squares problem) for the Attention Matching step.
​Question for the maintainers/ggerganov:
Does the team think this mathematical approach is feasible to implement within the current ggml architecture? What would be the recommended approach to implement or approximate the pseudoinverse operation in ggml efficiently across different backends?

### Analysis

According to the paper's benchmarks (tested on Qwen and Llama models), the proposed implementation maintains near-lossless generation quality on tasks like long-document analysis and coding, while drastically reducing the memory footprint. No actual llama.cpp profiling data is available yet since the ggml implementation doesn't exist.

### Relevant log output

```shell

```

**Comment by clayzermk1** [comment]
The paper author's code repo from the paper https://github.com/adamzweiger/compaction.

**Comment by Tedy50** [comment]
Not sure how similar it is, here is a proposal to get rid of KV cache entirely, promising to have a completely lossless implementation.
https://www.youtube.com/watch?v=TYgCRPCAFhE
If true, then you could have an unlimited context window on Raspberry Pi with no loss of performance at any context size

**Comment by Gavin0725** [comment]
> Not sure how similar it is, here is a proposal to get rid of KV cache entirely, promising to have a completely lossless implementation. https://www.youtube.com/watch?v=TYgCRPCAFhE If true, then you could have an unlimited context window on Raspberry Pi with no loss of performance at any context size

It doesn't actually get rid of the KV cache entirely. It still saves every token's Residual Stream. For example, in Gemma 3 12B, it uses about 7.68KB per token. While this is indeed much smaller than storing the full multi-layer KV cache, it is highly lossy because it only saves the last layer's residual stream.
​You cannot mathematically reverse-engineer the exact KV states of all the previous layers from just the final residual vector, because the data has already passed through numerous non-linear functions (like SwiGLU).
​On the other hand, if they tried to make it truly "lossless" by saving the residual for every layer, it would use the exact same amount of memory as the standard GQA KV cache, completely defeating the purpose of saving memory!
​Even if we accept the lossy compression, the compute overhead is the real bottleneck. Recomputing the KV matrices from the residual stream for every single historical token on the fly means the generation latency would be catastrophic. A Raspberry Pi would completely choke on the O(N) matrix multiplications for an "unlimited context".
​That's why the Fast KV Compaction paper is much more practical. It compresses the sequence length (N) directly. This not only shrinks the memory footprint drastically but also reduces the computation overhead, resulting in much faster generation speeds without the massive re-computation penalty.

**Comment by reversTeam** [comment]
The custom attention mask API in #21105 could be useful for attention pattern analysis during KV compaction research — it allows setting arbitrary masks and observing their effect on generation quality, which could help validate compaction decisions before committing them.
